File: Submariner/SBAppDelegate.m
//
//  SubmarinerAppDelegate.m
//  Submariner
//
//  Created by Rafaël Warnault on 06/06/11.
//
//  Copyright (c) 2011-2014, Rafaël Warnault
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  * Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  * Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  * Neither the name of the Read-Write.fr nor the names of its
//  contributors may be used to endorse or promote products derived from
//  this software without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
//  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
//  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
//  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
//  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#import "SBAppDelegate.h"
#import "SBPreferencesController.h"
#import "SBDatabaseController.h"
#import "SBTrack.h"
#import "SBPlayer.h"
#import "SBArtist.h"
#import "SBLibrary.h"
#import "SBAlbum.h"
// Additions
#import "NSManagedObjectContext+Fetch.h"

#import "Submariner-Swift.h"

@implementation SBAppDelegate


#pragma mark -
#pragma mark Singlton

+ (id)sharedInstance {
    // Cache since this can be called off main thread.
    // We get called by the main thread first, so this is OK.
    static SBAppDelegate *sharedApplication = nil;
    static dispatch_once_t token;
    dispatch_once(&token, ^{
        sharedApplication = (SBAppDelegate*)NSApplication.sharedApplication.delegate;
    });
    return sharedApplication;
}



#pragma mark -
#pragma mark LifeCycle


- (id)init {
    self = [super init];
    // XXX: Best place to initialize value transformers?
    SBRepeatModeTransformer *noneTrans = [[SBRepeatModeTransformer alloc] initWithMode: SBPlayerRepeatNo];
    [NSValueTransformer setValueTransformer: noneTrans forName: @"SBRepeatModeNoneTransformer"];
    SBRepeatModeTransformer *oneTrans = [[SBRepeatModeTransformer alloc] initWithMode: SBPlayerRepeatOne];
    [NSValueTransformer setValueTransformer: oneTrans forName: @"SBRepeatModeOneTransformer"];
    SBRepeatModeTransformer *allTrans = [[SBRepeatModeTransformer alloc] initWithMode: SBPlayerRepeatAll];
    [NSValueTransformer setValueTransformer: allTrans forName: @"SBRepeatModeAllTransformer"];
    
    // we have to do this because Swift
    SBTrackListLengthTransformer *lengthTrans = [[SBTrackListLengthTransformer alloc] init];
    [NSValueTransformer setValueTransformer: lengthTrans forName: @"SBTrackListLengthTransformer"];
    SBVolumeIconTransformer *volumeTrans = [[SBVolumeIconTransformer alloc] init];
    [NSValueTransformer setValueTransformer: volumeTrans forName: @"SBVolumeIconTransformer"];
    SBRepeatIconTransformer *repeatTrans = [[SBRepeatIconTransformer alloc] init];
    [NSValueTransformer setValueTransformer: repeatTrans forName: @"SBRepeatIconTransformer"];
    return self;
}




#pragma mark -
#pragma mark NSApplicationDelegate

- (void)applicationDidFinishLaunching:(NSNotification *)aNotification
{
    preferencesController = [[SBPreferencesController alloc] initWithManagedObjectContext:[self managedObjectContext]];
    databaseController = [[SBDatabaseController alloc] initWithManagedObjectContext:[self managedObjectContext]];
    
    [self zoomDatabaseWindow:nil];
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    
}

- (NSApplicationTerminateReply)applicationShouldTerminate:(NSApplication *)sender {
    
    // unplay all tracks before quitting
    NSError *error = nil;
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"(isPlaying == YES)"];
    NSArray *tracks = [[self managedObjectContext] fetchEntitiesNammed:@"Track" withPredicate:predicate error:&error];
    for(SBTrack *track in tracks) {
        [track setIsPlaying:[NSNumber numberWithBool:NO]];
    }
    
    // Save changes in the application's managed object context before the application terminates.
    if (!__managedObjectContext) {
        return NSTerminateNow;
    }
    
    if (![[self managedObjectContext] commitEditing]) {
        NSLog(@"%@:%@ unable to commit editing to terminate", [self class], NSStringFromSelector(_cmd));
        return NSTerminateCancel;
    }
    
    if (![[self managedObjectContext] hasChanges]) {
        return NSTerminateNow;
    }
    
    if (![[self managedObjectContext] save:&error]) {
        
        // Customize this code block to include application-specific recovery steps.              
        BOOL result = [sender presentError:error];
        if (result) {
            return NSTerminateCancel;
        }
        
        NSString *question = NSLocalizedString(@"Could not save changes while quitting. Quit anyway?", @"Quit without saves error question message");
        NSString *info = NSLocalizedString(@"Quitting now will lose any changes you have made since the last successful save", @"Quit without saves error question info");
        NSString *quitButton = NSLocalizedString(@"Quit anyway", @"Quit anyway button title");
        NSString *cancelButton = NSLocalizedString(@"Cancel", @"Cancel button title");
        NSAlert *alert = [[NSAlert alloc] init];
        [alert setMessageText:question];
        [alert setInformativeText:info];
        [alert addButtonWithTitle:quitButton];
        [alert addButtonWithTitle:cancelButton];
        
        NSInteger answer = [alert runModal];
        alert = nil;
        
        if (answer == NSAlertSecondButtonReturn) {
            return NSTerminateCancel;
        }
    }
    
    return NSTerminateNow;
}


- (BOOL) applicationShouldOpenUntitledFile:(NSApplication *)sender
{
    [self zoomDatabaseWindow:self];
    return NO;
}

- (BOOL)application:(NSApplication *)app openFile:(NSString *)filename {
    if(filename) {
        [databaseController openImportAlert:[databaseController window] files:[NSArray arrayWithObject:filename]];
        return YES;
    }
    return NO;
}

- (void)application:(NSApplication *)sender openFiles:(NSArray *)filenames {
    if(filenames && [filenames count] > 0)
        [databaseController openImportAlert:[databaseController window] files:filenames];
}

- (BOOL)application:(NSApplication *)app openFileWithoutUI:(NSString *)filename {
    if(filename) {
        [databaseController openImportAlert:[databaseController window] files:[NSArray arrayWithObject:filename]];
        return YES;
    }
    return NO; 
}







#pragma mark -
#pragma mark App Directory Management


- (NSURL *)applicationFilesDirectory {
    
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSURL *libraryURL = [[fileManager URLsForDirectory:NSLibraryDirectory inDomains:NSUserDomainMask] lastObject];
    return [libraryURL URLByAppendingPathComponent:@"Submariner"];
}

- (NSString *)musicDirectory {
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSString *path = [NSHomeDirectory() stringByAppendingPathComponent:@"Music/Submariner/Music"];
    if(![fileManager fileExistsAtPath:path]) [fileManager createDirectoryAtPath:path withIntermediateDirectories:YES attributes:nil error:nil];
    return path;
}

- (NSString *)coverDirectory {
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSString *path = [NSHomeDirectory() stringByAppendingPathComponent:@"Music/Submariner/Covers"];
    if(![fileManager fileExistsAtPath:path]) [fileManager createDirectoryAtPath:path withIntermediateDirectories:YES attributes:nil error:nil];
    return path;
}

- (NSString *)storeFileName {
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSString *path = [NSHomeDirectory() stringByAppendingPathComponent:@"Music/Submariner/Submariner Library.sqlite"];
    if(![fileManager fileExistsAtPath:path]) [fileManager createDirectoryAtPath:@"Music/Submariner/" withIntermediateDirectories:YES attributes:nil error:nil];
    return path;
}





#pragma mark -
#pragma mark IBAction

- (IBAction) saveAction:(id)sender {
    
    NSError *error = nil;
    
    if([[self managedObjectContext] hasChanges]) {
        NSLog(@"save : hasChange");
        if (![[self managedObjectContext] commitEditing]) {
            NSLog(@"%@:%@ unable to commit editing before saving", [self class], NSStringFromSelector(_cmd));
        }
        
        if (![[self managedObjectContext] save:&error]) {
            [[NSApplication sharedApplication] presentError:error];
        }
    }
}

- (IBAction)zoomDatabaseWindow:(id)sender {
    
    NSWindow *window = [databaseController window];
    
    // zoom the window, useless ?
//    NSRect rect = NSMakeRect([window frame].origin.x+[window frame].size.width/2,
//                             [window frame].origin.y+[window frame].size.height/2,
//                             10,
//                             10);    
//    
//    [window zoomOnFromRect:rect];
    
    [window makeKeyAndOrderFront:sender];
}


- (IBAction)openPreferences:(id)sender {
    [preferencesController showWindow:sender];
}

- (IBAction)openDatabase:(id)sender {
    [databaseController showWindow:sender];
}

- (IBAction)openAudioFiles:(id)sender {
    [databaseController openAudioFiles:sender];
}

- (IBAction)newPlaylist:(id)sender {
    [databaseController addPlaylist:sender];
}

- (IBAction)addPlaylistToCurrentServer:(id)sender {
    [databaseController addPlaylistToCurrentServer:sender];
}

- (IBAction)newServer:(id)sender {
    [databaseController addServer:sender];
}

- (IBAction)toogleTracklist:(id)sender {
    [databaseController toggleTrackList:sender];
}

- (IBAction)toggleServerUsers:(id)sender {
    [databaseController toggleServerUsers:sender];
}

- (IBAction)playPause:(id)sender {
    [databaseController playPause:sender];
}

- (IBAction)stop:(id)sender {
    [databaseController stop:sender];
}

- (IBAction)nextTrack:(id)sender {
    [databaseController nextTrack:sender];
}

- (IBAction)previousTrack:(id)sender {
    [databaseController previousTrack:sender];
}

- (IBAction)repeatNone:(id)sender {
    [databaseController repeatNone: sender];
}

- (IBAction)repeatOne:(id)sender {
    [databaseController repeatOne: sender];
}

- (IBAction)repeatAll:(id)sender {
    [databaseController repeatAll: sender];
}

- (IBAction)repeatModeCycle:(id)sender {
    [databaseController repeat: sender];
}

- (IBAction)toggleShuffle:(id)sender {
    [databaseController shuffle: sender];
}

- (IBAction)rewind:(id)sender {
    [databaseController rewind: sender];
}

- (IBAction)fastForward:(id)sender {
    [databaseController fastForward: sender];
}

- (IBAction)setMuteOn:(id)sender {
    [databaseController setMuteOn: sender];
}

- (IBAction)volumeUp:(id)sender {
    [databaseController volumeUp: sender];
}

- (IBAction)volumeDown:(id)sender {
    [databaseController volumeDown: sender];
}

- (IBAction)search:(id)sender {
    [databaseController search:sender];
}

- (IBAction)showIndices:(id)sender {
    [databaseController showIndices:sender];
}

- (IBAction)showAlbums:(id)sender {
    [databaseController showAlbums:sender];
}

- (IBAction)showPodcasts:(id)sender {
    [databaseController showPodcasts:sender];
}

- (IBAction)cleanTracklist:(id)sender {
    [databaseController cleanTracklist:sender];
}

- (IBAction)showWebsite:(id)sender {
    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:@"http://submarinerapp.com/"]];
}

- (IBAction)reloadCurrentServer:(id)sender {
    [databaseController reloadCurrentServer:sender];
}

- (IBAction)openCurrentServerHomePage:(id)sender {
    [databaseController openCurrentServerHomePage:sender];
}

- (IBAction)goToCurrentTrack:(id)sender {
    [databaseController goToCurrentTrack:sender];
}

- (IBAction)renameItem:(id)sender {
    [databaseController renameItem:sender];
}

- (IBAction)configureCurrentServer:(id)sender {
    [databaseController configureCurrentServer:sender];
}


#pragma mark -
#pragma mark Core Data Support


/**
    Creates if necessary and returns the managed object model for the application.
 */
- (NSManagedObjectModel *)managedObjectModel {
    if (__managedObjectModel) {
        return __managedObjectModel;
    }
	
    NSURL *modelURL = [[NSBundle mainBundle] URLForResource:@"Submariner" withExtension:@"momd"];
    __managedObjectModel = [[NSManagedObjectModel alloc] initWithContentsOfURL:modelURL];    
    return __managedObjectModel;
}

/**
    Returns the persistent store coordinator for the application. This implementation creates and return a coordinator, having added the store for the application to it. (The directory for the store is created, if necessary.)
 */
- (NSPersistentStoreCoordinator *) persistentStoreCoordinator {
    if (__persistentStoreCoordinator) {
        return __persistentStoreCoordinator;
    }

    NSManagedObjectModel *mom = [self managedObjectModel];
    if (!mom) {
        NSLog(@"%@:%@ No model to generate a store from", [self class], NSStringFromSelector(_cmd));
        return nil;
    }
    
    NSError *error = nil;
    
    NSURL *oldUrl = [[self applicationFilesDirectory] URLByAppendingPathComponent:@"Submariner.storedata"];
    NSURL *url = [NSURL fileURLWithPath: [self storeFileName]];
    __persistentStoreCoordinator = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:mom];
    NSDictionary *storeOpts = @{
        NSInferMappingModelAutomaticallyOption: @YES,
        NSMigratePersistentStoresAutomaticallyOption: @YES
    };
    if (![[NSFileManager defaultManager] fileExistsAtPath: [url path]] && [[NSFileManager defaultManager] fileExistsAtPath: [oldUrl path]]) {
        NSPersistentStore *oldStore = [__persistentStoreCoordinator addPersistentStoreWithType:NSXMLStoreType configuration:nil URL:oldUrl options: storeOpts error:&error];
        if (oldStore == nil) {
            [[NSApplication sharedApplication] presentError:error];
            __persistentStoreCoordinator = nil;
            return nil;
        }
        NSPersistentStore *newStore = [__persistentStoreCoordinator migratePersistentStore:oldStore toURL:url options:storeOpts withType:NSSQLiteStoreType error:&error];
        if (newStore == nil) {
            [[NSApplication sharedApplication] presentError:error];
            __persistentStoreCoordinator = nil;
            return nil;
        }
        // old store is removed from coordinator
        oldStore = nil;
        // XXX: Remove/rename the old file?
    } else {
        // Only use the SQLite store.
        if (![__persistentStoreCoordinator addPersistentStoreWithType:NSSQLiteStoreType configuration:nil URL:url options: storeOpts error:&error]) {
            [[NSApplication sharedApplication] presentError:error];
            __persistentStoreCoordinator = nil;
            return nil;
        }
    }

    return __persistentStoreCoordinator;
}

/**
    Returns the managed object context for the application (which is already
    bound to the persistent store coordinator for the application.) 
 */
- (NSManagedObjectContext *) managedObjectContext {
    if (__managedObjectContext) {
        return __managedObjectContext;
    }

    NSPersistentStoreCoordinator *coordinator = [self persistentStoreCoordinator];
    if (!coordinator) {
        NSMutableDictionary *dict = [NSMutableDictionary dictionary];
        [dict setValue:@"Failed to initialize the store" forKey:NSLocalizedDescriptionKey];
        [dict setValue:@"There was an error building up the data file." forKey:NSLocalizedFailureReasonErrorKey];
        NSError *error = [NSError errorWithDomain:@"YOUR_ERROR_DOMAIN" code:9999 userInfo:dict];
        [[NSApplication sharedApplication] presentError:error];
        return nil;
    }
    __managedObjectContext = [[NSManagedObjectContext alloc] initWithConcurrencyType: NSPrivateQueueConcurrencyType];
    [__managedObjectContext setPersistentStoreCoordinator:coordinator];

    return __managedObjectContext;
}



- (NSUndoManager *)windowWillReturnUndoManager:(NSWindow *)window {
    return [[self managedObjectContext] undoManager];
}



#pragma mark -
#pragma mark UI Validator

- (BOOL)validateUserInterfaceItem: (id<NSValidatedUserInterfaceItem>) item {
    return [databaseController validateUserInterfaceItem: item];
}




@end

File: Submariner/SBAppDelegate.swift
//
//  SBAppDelegate.swift
//  Submariner
//
//  Created by Calvin Buckley on 2023-06-17.
//  Copyright © 2023 Submariner Developers. All rights reserved.
//

import Cocoa
import os

fileprivate let logger = Logger(subsystem: Bundle.main.bundleIdentifier!, category: "SBAppDelegate")

@objc(SBAppDelegate) class SBAppDelegate: NSObject, NSApplicationDelegate, NSMenuDelegate, NSUserInterfaceValidations {
    
    // #MARK: - Singleton
    
    private static var _sharedInstance = SBAppDelegate()
    
    // FIXME: Make var
    @objc static func sharedInstance() -> SBAppDelegate {
        return _sharedInstance
    }
    
    // #MARK: - Initialization
    
    let databaseController: SBDatabaseController
    let preferencesController: SBPreferencesController
    
    override init() {
        // #MARK: Init User Defaults
        let defaults: [String: Any] = [
            "clientIdentifier": "submariner",
            "apiVersion": "1.15.0",
            "playerBehavior": NSNumber(value: 1),
            "playerVolume": NSNumber(value: 0.5),
            "repeatMode": NSNumber(value: SBPlayer.RepeatMode.no.rawValue),
            "shuffle": NSNumber(value: false),
            "enableCacheStreaming": NSNumber(value: true),
            "autoRefreshNowPlaying": NSNumber(value: false),
            "coverSize": NSNumber(value: 0.75),
            "maxBitRate": NSNumber(value: 0),
            "MaxCoverSize": NSNumber(value: 300),
            "scrobbleToServer": NSNumber(value: true),
            "deleteAfterPlay": NSNumber(value: false),
            "SkipIncrement": NSNumber(value: 5.0)
        ]
        UserDefaults.standard.register(defaults: defaults)
        
        // #MARK: Init Value Transformers
        // other NSVTs are found by objc runtime by name
        let noneTrans = SBRepeatModeTransformer(mode: .no)
        let noneTransName = NSValueTransformerName(rawValue: "SBRepeatModeNoneTransformer")
        ValueTransformer.setValueTransformer(noneTrans, forName: noneTransName)
        let oneTrans = SBRepeatModeTransformer(mode: .one)
        let oneTransName = NSValueTransformerName(rawValue: "SBRepeatModeOneTransformer")
        ValueTransformer.setValueTransformer(oneTrans, forName: oneTransName)
        let allTrans = SBRepeatModeTransformer(mode: .all)
        let allTransName = NSValueTransformerName(rawValue: "SBRepeatModeAllTransformer")
        ValueTransformer.setValueTransformer(allTrans, forName: allTransName)
        
        // #MARK: Init Core Data (managed object model)
        let modelURL = Bundle.main.url(forResource: "Submariner", withExtension: "momd")!
        self.managedObjectModel = NSManagedObjectModel(contentsOf: modelURL)!
        
        // #MARK: Init Core Data (persistent store coordinator)
        self.persistentStoreCoordinator = NSPersistentStoreCoordinator(managedObjectModel: self.managedObjectModel)
        let storeOpts = [
            NSInferMappingModelAutomaticallyOption: true,
            NSMigratePersistentStoresAutomaticallyOption: true
        ]
        // migrate legacy stores when possible; legacyStoreFile
        // XXX: Convert to the nicer Swift API in macOS 12+ when we drop 11 support
        // XXX: Error handling sufficient here?
        let oldURL = SBAppDelegate.legacyStoreFileName
        let newURL = SBAppDelegate.storeFileName
        if FileManager.default.fileExists(atPath: oldURL.path) && !FileManager.default.fileExists(atPath: newURL.path) {
            let oldStore = try! self.persistentStoreCoordinator.addPersistentStore(ofType: NSXMLStoreType,
                                                                                   configurationName: nil,
                                                                                   at: oldURL,
                                                                                   options: storeOpts)
            try! self.persistentStoreCoordinator.migratePersistentStore(oldStore,
                                                                        to: newURL,
                                                                        options: storeOpts,
                                                                        withType: NSSQLiteStoreType)
        } else {
            // usual path, we aren't converting, but just using modern store
            try! self.persistentStoreCoordinator.addPersistentStore(ofType: NSSQLiteStoreType,
                                                                    configurationName: nil,
                                                                    at: newURL,
                                                                    options: storeOpts)
        }
        
        // #MARK: Init Core Data (managed object store)
        // must be main queue for SwiftUI
        self.managedObjectContext = NSManagedObjectContext(concurrencyType: .mainQueueConcurrencyType)
        self.managedObjectContext.persistentStoreCoordinator = self.persistentStoreCoordinator
        
        // #MARK: Init Window Controllers
        self.databaseController = SBDatabaseController(managedObjectContext: self.managedObjectContext)
        self.preferencesController = SBPreferencesController(managedObjectContext: self.managedObjectContext)
    }
    
    // #MARK: - NSApplicationDelegate
    
    func applicationDidFinishLaunching(_ notification: Notification) {
        zoomDatabaseWindow(self)
    }
    
    func applicationShouldTerminate(_ sender: NSApplication) -> NSApplication.TerminateReply {
        SBPlayer.sharedInstance().stop()
        
        // If we have database corruption, we're screwed anyways and shouldn't put the user in an infinite loop.
        if !managedObjectContext.commitEditing() {
            let alert = NSAlert()
            alert.alertStyle = .critical
            alert.messageText = "Failed to Commit Changes"
            alert.informativeText = "Submariner failed to commit changes to the local database while exiting."
            alert.runModal()
            return .terminateNow
        }
        
        if !managedObjectContext.hasChanges {
            return .terminateNow
        }
        
        do {
            try managedObjectContext.save()
        } catch {
            if NSApplication.shared.presentError(error) {
                return .terminateCancel
            }
            let alert = NSAlert()
            alert.alertStyle = .critical
            alert.messageText = "Could not save changes while quitting. Quit anyway?"
            alert.informativeText = "Quitting now will lose any changes you have made since the last successful save."
            alert.addButton(withTitle: "Quit")
            alert.addButton(withTitle: "Cancel")
            if alert.runModal() == .cancel {
                return .terminateCancel
            }
        }
        
        return .terminateNow
    }
    
    func applicationOpenUntitledFile(_ sender: NSApplication) -> Bool {
        zoomDatabaseWindow(self)
        return false
    }
    
    func application(_ sender: NSApplication, openFile filename: String) -> Bool {
        databaseController.openImportAlert(databaseController.window, files: [filename])
        return true
    }
    
    func application(_ sender: Any, openFileWithoutUI filename: String) -> Bool {
        databaseController.openImportAlert(databaseController.window, files: [filename])
        return true
    }
    
    func application(_ sender: NSApplication, openFiles filenames: [String]) {
        databaseController.openImportAlert(databaseController.window, files: filenames)
    }
    
    // #MARK: - Application Files/Directories
    
    static var legacyStoreFileName: URL {
        let libraryURL = FileManager.default.urls(for: .libraryDirectory, in: .userDomainMask).last!
        let storeURL = libraryURL.appendingPathComponent("Submariner.storedata")
        return storeURL
    }
    
    @objc static var musicDirectory: URL {
        let path = FileManager.default.urls(for: .musicDirectory, in: .userDomainMask).last!.appendingPathComponent("Submariner/Music")
        if !FileManager.default.fileExists(atPath: path.path) {
            try! FileManager.default.createDirectory(at: path, withIntermediateDirectories: true)
        }
        return path
    }
    
    @objc static var coverDirectory: URL {
        let path = FileManager.default.urls(for: .musicDirectory, in: .userDomainMask).last!.appendingPathComponent("Submariner/Covers")
        if !FileManager.default.fileExists(atPath: path.path) {
            try! FileManager.default.createDirectory(at: path, withIntermediateDirectories: true)
        }
        return path
    }
    
    static var storeFileName: URL {
        let baseURL = FileManager.default.urls(for: .musicDirectory, in: .userDomainMask).last!.appendingPathComponent("Submariner")
        let path = FileManager.default.urls(for: .musicDirectory, in: .userDomainMask).last!.appendingPathComponent("Submariner/Submariner Library.sqlite")
        if !FileManager.default.fileExists(atPath: baseURL.path) {
            try! FileManager.default.createDirectory(at: baseURL, withIntermediateDirectories: true)
        }
        return path
    }
    
    // #MARK: - Outlets
    
    @IBAction func showWebsite(_ sender: Any?) {
        NSWorkspace.shared.open(URL(string: "https://submarinerapp.com")!)
    }
    
    @IBAction func saveAction(_ sender: Any?) {
        if managedObjectContext.hasChanges {
            do {
                managedObjectContext.commitEditing()
                try managedObjectContext.save()
            } catch {
                NSApplication.shared.presentError(error)
            }
        }
    }
    
    @IBAction func zoomDatabaseWindow(_ sender: Any?) {
        databaseController.window?.makeKeyAndOrderFront(sender)
    }
    
    @IBAction func openPreferences(_ sender: Any?) {
        preferencesController.showWindow(sender)
    }
    
    @IBAction func openDatabase(_ sender: Any?) {
        databaseController.showWindow(sender)
    }
    
    @IBAction func openAudioFiles(_ sender: Any?) {
        databaseController.openAudioFiles(sender)
    }
    
    @IBAction func newPlaylist(_ sender: Any?) {
        databaseController.addPlaylist(sender)
    }
    
    @IBAction func addPlaylistToCurrentServer(_ sender: Any?) {
        databaseController.addPlaylistToCurrentServer(sender)
    }
    
    @IBAction func newServer(_ sender: Any?) {
        databaseController.addServer(sender)
    }
    
    @IBAction func toogleTracklist(_ sender: Any?) {
        databaseController.toggleTrackList(sender)
    }
    
    @IBAction func toggleServerUsers(_ sender: Any?) {
        databaseController.toggleServerUsers(sender)
    }
    
    @IBAction func playPause(_ sender: Any?) {
        databaseController.playPause(sender)
    }
    
    @IBAction func stop(_ sender: Any?) {
        databaseController.stop(sender)
    }
    
    @IBAction func nextTrack(_ sender: Any?) {
        databaseController.nextTrack(sender)
    }
    
    @IBAction func previousTrack(_ sender: Any?) {
        databaseController.previousTrack(sender)
    }
    
    @IBAction func repeatNone(_ sender: Any?) {
        databaseController.repeatNone(sender)
    }
    
    @IBAction func repeatOne(_ sender: Any?) {
        databaseController.repeatOne(sender)
    }
    
    @IBAction func repeatAll(_ sender: Any?) {
        databaseController.repeatAll(sender)
    }
    
    @IBAction func repeatModeCycle(_ sender: Any?) {
        databaseController.repeat(sender)
    }
    
    @IBAction func toggleShuffle(_ sender: Any?) {
        databaseController.shuffle(sender)
    }
    
    @IBAction func rewind(_ sender: Any?) {
        databaseController.rewind(sender)
    }
    
    @IBAction func fastForward(_ sender: Any?) {
        databaseController.fastForward(sender)
    }
    
    @IBAction func setMuteOn(_ sender: Any?) {
        databaseController.setMuteOn(sender)
    }
    
    @IBAction func volumeUp(_ sender: Any?) {
        databaseController.volumeUp(sender)
    }
    
    @IBAction func volumeDown(_ sender: Any?) {
        databaseController.volumeDown(sender)
    }
    
    @IBAction func search(_ sender: Any?) {
        databaseController.search(sender)
    }
    
    @IBAction func showIndices(_ sender: Any?) {
        databaseController.showIndices(sender)
    }
    
    @IBAction func showAlbums(_ sender: Any?) {
        databaseController.showAlbums(sender)
    }
    
    @IBAction func showPodcasts(_ sender: Any?) {
        databaseController.showPodcasts(sender)
    }
    
    @IBAction func cleanTracklist(_ sender: Any?) {
        databaseController.cleanTracklist(sender)
    }
    
    @IBAction func reloadCurrentServer(_ sender: Any?) {
        databaseController.reloadCurrentServer(sender)
    }
    
    @IBAction func openCurrentServerHomePage(_ sender: Any?) {
        databaseController.openCurrentServerHomePage(sender)
    }
    
    @IBAction func goToCurrentTrack(_ sender: Any?) {
        databaseController.goToCurrentTrack(sender)
    }
    
    @IBAction func renameItem(_ sender: Any?) {
        databaseController.renameItem(sender)
    }
    
    @IBAction func configureCurrentServer(_ sender: Any?) {
        databaseController.configureCurrentServer(sender)
    }
    
    @IBAction func scanCurrentLibrary(_ sender: Any?) {
        databaseController.scanCurrentLibrary(sender)
    }
    
    @IBAction func purgeLocalLibrary(_ sender: Any?) {
        let operation = SBLibraryPurgeOperation(managedObjectContext: managedObjectContext)
        OperationQueue.sharedServerQueue.addOperation(operation)
    }
    
    // #MARK: - Core Data
    
    @objc let managedObjectModel: NSManagedObjectModel
    @objc let persistentStoreCoordinator: NSPersistentStoreCoordinator
    @objc let managedObjectContext: NSManagedObjectContext
    
    // #MARK: - UI Validation
    
    func validateUserInterfaceItem(_ item: NSValidatedUserInterfaceItem) -> Bool {
        return databaseController.validate(item)
    }
}
